dhl_sdk/_input_processing.py

```diff
@@ -0,0 +1,718 @@
+"""This module contains utility functions for data validation and formatting in the SDK
+"""
+
+# pylint: disable=too-few-public-methods
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+import math
+from typing import Optional, Protocol, Union
+
+import numpy as np
+
+
+from dhl_sdk._utils import (
+    Predictions,
+    PredictResponse,
+    validate_list_elements,
+)
+from dhl_sdk.exceptions import (
+    InvalidSpectraException,
+    InvalidInputsException,
+    InvalidStepsException,
+    InvalidTimestampsException,
+)
+from dhl_sdk._spectra_utils import (
+    SpectraData,
+    SpectraModel,
+    _validate_spectra_format,
+    _convert_to_request,
+)
+
+
+class Group(Protocol):
+    # pylint: disable=missing-class-docstring
+    # pylint: disable=missing-function-docstring
+    @property
+    def code(self) -> str: ...
+
+
+class Variable(Protocol):
+    # pylint: disable=missing-class-docstring
+    # pylint: disable=missing-function-docstring
+    @property
+    def group(self) -> Group: ...
+
+    @property
+    def code(self) -> str: ...
+
+    @property
+    def id(self) -> str: ...
+
+    def matches_key(self, key: str) -> bool: ...
+
+
+class Dataset(Protocol):
+    # pylint: disable=missing-class-docstring
+    # pylint: disable=missing-function-docstring
+    @property
+    def variables(self) -> list[Variable]: ...
+
+
+class Model(Protocol):
+    # pylint: disable=missing-class-docstring
+    # pylint: disable=missing-function-docstring
+    @property
+    def dataset(self) -> Dataset: ...
+
+
+# Input Preprocessors (Validation and Formatting)
+class Preprocessor(ABC):
+    """Interface for preprocessing"""
+
+    @abstractmethod
+    def validate(self) -> bool:
+        """Validate the input data"""
+
+    @abstractmethod
+    def format(self) -> list[dict]:
+        """Format the input data to a list of JSON requests"""
+
+
+@dataclass
+class SpectraPreprocessor(Preprocessor):
+    """Implementation of the preprocessor for Spectrum models"""
+
+    spectra: SpectraData
+    inputs: Optional[dict]
+    model: SpectraModel
+
+    def validate(self) -> bool:
+        """
+        This function validates and makes the necessary formating of the spectra
+        and inputs used for prediction. It performs the following validations:
+
+        - Validates if spectra is empty
+        - Validates if the number of wavelengths in spectra matches
+            the number of wavelengths in the model
+        - Validates if the spectra contains None values
```

Comment:
grammar:
- Validates if the spectra are empty
- Validates if the number of wavelengths in the spectra match the number of wavelengths in the model
- Validates if the spectra contain None (missing) values